### ISSUE TITLE
nc-complex and h5netcdf

### ISSUE THREAD

**Issue by kmuehlbauer** [PlasmaFAIR/nc-complex#12]
Great work to get complex variable into netcdf4-python, @ZedThree.

I'm maintainer of [h5netcdf](https://github.com/h5netcdf/h5netcdf), reading and writing netCDF4 files via `h5py`. As `h5py`/`h5netcdf` are capable of reading/writing complex data as compound types, same compound type you are using here:

```
DATATYPE "_PFNC_DOUBLE_COMPLEX_TYPE" H5T_COMPOUND {
      H5T_IEEE_F64LE "r";
      H5T_IEEE_F64LE "i";
   }
   DATASET "data" {
      DATATYPE  H5T_COMPOUND {
         H5T_IEEE_F64LE "r";
         H5T_IEEE_F64LE "i";
      }
   ...
}
```

I'm trying to implement committing the necessary named type to the file so that netcdf4-python/nc-complex can directly read it. Unfortunately this does not seem to be that straightforward. Something weird is going on probably with `DIMENSION_LIST`-entry of the `data`-variable and/or `REFERENCE_LIST` entry of the `x`-dimension.

I'd appreciate if you could share some insight on how to trick the two files to behave the same. `test_complex1` is created with latest `netcdf4-python/nc_complex` and `test_complex2` is created with `h5netcdf`.

Thanks a bunch!

[test_complex2.tar.gz](https://github.com/PlasmaFAIR/nc-complex/files/13960398/test_complex2.tar.gz)
[test_complex1.tar.gz](https://github.com/PlasmaFAIR/nc-complex/files/13960399/test_complex1.tar.gz)

**Comment by ZedThree** [comment]
Hi @kmuehlbauer, this would be really great for portability! The next release of netCDF-C will also have support for the unnamed/non-committed types that `h5py`/`h5netcdf` can do already, but committing the type will mean users will be able to use older netCDF versions too.

If I use `ncdump` from the current `main` of netcdf-C which has this feature, I can see that the variable is still using an unnamed type:

```
netcdf test_complex2 {
types:
  compound _AnonymousCompound1 {
    double r ;
    double i ;
  }; // _AnonymousCompound1
  compound _PFNC_DOUBLE_COMPLEX_TYPE {
    double r ;
    double i ;
  }; // _PFNC_DOUBLE_COMPLEX_TYPE
dimensions:
        x = 5 ;
variables:
        _AnonymousCompound1 data(x) ;
data:

 data = {0, 0}, {1, 0}, {0, 1}, {1, 1}, {0.25, 0.75} ;
}
```

I think the reason you're not seeing a difference from `h5dump` is because it copies the definition of the compound type into the variable (when dumping at least, not sure about actual storage), rather than reporting it as a reference. This makes the files look identical, when really the `h5netcdf` version has an extra (anonymous) type definition.

This is probably something best fixed at the `h5py` level, rather than `h5netcdf` -- I've previously opened an issue about it (https://github.com/h5py/h5py/issues/2291). This would also improve compatibility for booleans and enums at the same time.

**Comment by kmuehlbauer** [comment]
Thanks @ZedThree for the fast response. 

I'm a bit puzzled now, as this works fine if I create the compound type the usual way:

```python
with netCDF4.Dataset("complex1.nc","w") as ds:
    size = 5
    datac = np.exp(1j*(1.+np.linspace(0, np.pi, size)))
    complex128 = np.dtype([("real",np.float64),("imag",np.float64)])
    complex128_t = ds.createCompoundType(complex128, "complex128")
    ds.createDimension("x", None)
    v = ds.createVariable("complex_var", complex128_t, ("x",))
    data = np.empty(size, complex128)
    data["real"] = datac.real
    data["imag"] = datac.imag
    v[:] = data
```

The same code works for h5netcdf's legacyapi (local branch, not yet released). The files are looking good with `ncdump`. So I really do not understand why I'm not able to mimic that to be used by `nc-complex`. Good chance, that `h5py` is doing some special handling here. Digging deeper now. 

[complex1.tar.gz](https://github.com/PlasmaFAIR/nc-complex/files/13962250/complex1.tar.gz)
[complex2.tar.gz](https://github.com/PlasmaFAIR/nc-complex/files/13962251/complex2.tar.gz)

**Comment by kmuehlbauer** [comment]
@ZedThree I think I might be doing something wrong on my side here. I'll come back with a solution hopefully later. No need for immediate action from your side. Thanks!

**Comment by ZedThree** [comment]
Please could you post or link to the code that isn't working? I suspect that the issue is that when you pass complex numbers off to `h5py` it's still automatically making an anonymous type. Relevant bit in `h5py` is [this bit in h5t.pyx](https://github.com/h5py/h5py/blob/a08b7014aa5eeaeee9dbd4f30345cb8be1b11ce1/h5py/h5t.pyx#L1581-L1583).

I think it "just" needs a call to [`commit`](https://api.h5py.org/h5t.html#h5py.h5t.TypeID.commit) somewhere. Although you need a specific group to commit to, as well as some logic for checking if it already exists.

**Comment by kmuehlbauer** [comment]
@ZedThree Sorry for the noise, I found the issue.

When creating the variable, I did not use the right type (just the plain numpy type). So h5netcdf/h5py created the unlinked transient type. I'm now checking the existing compound types for complex types and use those when creating the variable. That seems to work because ncdump is exporting same values. 

I'm shaping the code now and have a PR soon ready over at h5netcdf. I can ping you then from there, if you are interested. Thanks again!
